engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java

```diff
@@ -154,29 +168,81 @@ private boolean typeCanOverride(MethodRootNode node, EnsoContext ctx) {
-          Object self,
-          Object[] arguments,
-          @CachedLibrary(limit = "10") TypesLibrary typesLibrary,
-          @Cached MethodResolverNode methodResolverNode) {
+      VirtualFrame frame,
+      State state,
+      UnresolvedSymbol symbol,
+      Object self,
+      Object[] arguments,
+      @CachedLibrary(limit = "10") TypesLibrary typesLibrary,
+      @Cached MethodResolverNode methodResolverNode) {
     Type selfTpe = typesLibrary.getType(self);
     Function function = resolveFunction(symbol, selfTpe, methodResolverNode);
     if (function == null) {
       throw new PanicException(
-              EnsoContext.get(this).getBuiltins().error().makeNoSuchMethod(self, symbol), this);
+          EnsoContext.get(this).getBuiltins().error().makeNoSuchMethod(self, symbol), this);
+    }
+    var resolvedFuncArgCount = function.getSchema().getArgumentsCount();
+    CallArgumentInfo[] invokeFuncSchema = invokeFunctionNode.getSchema();
+    // Check if this is a static method call on Any
+    if (isAnyEigenType(selfTpe) && arguments.length == resolvedFuncArgCount - 1) {
+      // function is a static method on Any, so the first two arguments in `invokeFuncSchema`
+      // represent self arguments.
+      boolean selfArgSpecified = false;
+      if (invokeFuncSchema.length > 1) {
+        selfArgSpecified =
+            invokeFuncSchema[1].getName() != null && invokeFuncSchema[1].getName().equals("self");
+      }
+
+      if (selfArgSpecified) {
+        // If there is a self named argument in the method call, we fall back to the old
+        // behavior - there will be no prepended Any.type self argument
+        assert arguments.length == invokeFunctionNode.getSchema().length;
+        return invokeFunctionNode.execute(function, frame, state, arguments);
+      }
+
+      Object[] argsWithPrependedSelf = new Object[arguments.length + 1];
+      argsWithPrependedSelf[0] = EnsoContext.get(this).getBuiltins().any().getEigentype();
+      System.arraycopy(arguments, 0, argsWithPrependedSelf, 1, arguments.length);
+
+      if (invokeAnyFunctionNode == null) {
+        CompilerDirectives.transferToInterpreter();
+        assert resolvedFuncArgCount >= 2
+            : "Resolved function should be on Any.type, therefore, should have at least two self"
+                + " arguments";
+        // Prepend self=Any to the arguments and shift
+        assert resolvedFuncArgCount - 1 <= invokeFuncSchema.length
+            && invokeFuncSchema.length <= resolvedFuncArgCount;
+        assert arguments.length == invokeFuncSchema.length;
+
+        CallArgumentInfo[] newInvokeFuncSchema = new CallArgumentInfo[resolvedFuncArgCount];
+        newInvokeFuncSchema[0] = new CallArgumentInfo("self");
+        newInvokeFuncSchema[1] = new CallArgumentInfo("self");
+        int toCopyLen = Math.min(newInvokeFuncSchema.length - 2, invokeFuncSchema.length);
+        System.arraycopy(invokeFuncSchema, 0, newInvokeFuncSchema, 2, toCopyLen);
```

Comment:
Last time I tried to do similar manipulation of arguments I was (rightfully) reminded that this approach is really bad and sounds like a hack, not to mention perf penalty.

Reply:
I don't think either, this is an elegant solution. It's the only one that I was able to come up with. Why do you think there will be any perf penalty? The penalty is paid just in the very first invocation of such a special static method invocation on `Any`, which will construct `invokeAnyFunctionNode`. Once this node is constructed, all the subsequent calls should have negligible perf penalty.

Reply:
Nevertheless, I will be more than happy for any other suggestions